File: solutions/python/Maximum Level Sum of a Binary Tree.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        queue=[root]
        max_level=0
        level=0
        max_sum=root.val
        while queue:
            size=len(queue)
            _sum=0
            for i in range(size):
                node=queue.pop(0)
                _sum=_sum+node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            level=level+1
            if level==1 and _sum>=max_sum:
                max_level=level
                max_sum=_sum
            elif _sum>max_sum:
                max_sum=_sum
                max_level=level
        return max_level

File: solutions/Typescript/Maximum Level Sum of a Binary Tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function maxLevelSum(root: TreeNode | null): number {
  if (root === null) return 0;
  let queue: TreeNode[] = [root];
  let max_sum: number = root.val;
  let level: number = 0;
  let max_level: number = 0;
  while (queue.length > 0) {
    let size: number = queue.length;
    let sum: number = 0;
    for (let i: number = 0; i < size; i++) {
      let node: TreeNode = queue.shift();
      sum = sum + node.val;
      if (node.left !== null) {
        queue.push(node.left);
      }
      if (node.right !== null) {
        queue.push(node.right);
      }
    }
    level = level + 1;
    if (level === 1 && sum >= max_sum) {
      max_sum = sum;
      max_level = level;
    } else if (sum > max_sum) {
      max_sum = sum;
      max_level = level;
    }
  }
  return max_level;
}
